app/controllers/samples_reports_controller.rb

```diff
@@ -135,6 +137,29 @@ def update_threshold
     render json: { threshold: threshold, confusion_matrix: confusion_matrix }
   end
 
+  def download_nih_tables
```

Comment:
You know, I'm not a big fan of non-CRUD methods, to be honest. But I've got a proposal that I think you might find interesting

```
class NIHTablesController < ApplicationController
  before_action :set_samples_report

  def show
    samples_report = SamplesReport.find(params[:samples_report_id])
    return unless authorize_resource(samples_report, READ_SAMPLES_REPORT)
    
    purpose = samples_report.samples[0].box.purpose
    zip_file = create_zip_file(samples_report.name)

    if purpose == "LOD"
      add_lod_table(zip_file)
    elsif purpose == "Challenge"
      add_challenge_table(zip_file)
    end

    zip_file.close

    send_zip_file(zip_file, samples_report.name)
  end

  private

  def create_zip_file(filename)
    zip_file = Tempfile.new("#{filename}_nih_tables.zip")
    Zip::File.open(zip_file.path, Zip::File::CREATE) do |zip|
      zip.add("Instructions.txt", Rails.root.join('public/templates/Instructions.txt'))
    end
    zip_file
  end

  def add_lod_table(zip_file)
    # TODO: Add the LOD table to the zip file
  end

  def add_challenge_table(zip_file)
    # TODO: Add the Challenge table to the zip file
  end

  def send_zip_file(zip_file, filename)
    send_file zip_file.path, type: 'application/zip', filename: "#{filename}_nih_tables.zip"
  end
end

```

routes

```
 resources :samples_reports, only: [] do
    member do
      get 'nih_tables', to: 'nih_tables#show'
    end
  end
```

Reply:
change my mind. CREATE is better than Show you're creating a new download each time, but it comes with its own set of trade-offs 

`resources :nih_tables, only: [:create]`

Reply:
Yes, here it makes sense to create another controller!

You may even nest it under the :samples_reports resource:

```ruby
resources :samples_reports do
  resources :nih_tables, only: [:create]
end
```

So you get an explicit route name, and the explicit `:samples_report_id` param in your controller, instead of the plain `:id` that could give the impression that "nih tables" is a resource.

```ruby
samples_reports_nih_tables_path(samples_report)
# => /samples_reports/:samples_report_id/nih_tables
```

Reply:
Note: it might not be easy to use `#create` instead of `#show` because we link from a React component where it might be more complex than necessary to create a form with the CSRF token :/

Reply:
By nesting I'm asked for a `respond_to` a format and the route becomes a post... But I'm doing a `GET` from js, not from react. Like it is, is working ok with the `#create` method (and also with the `#show` method).

`/views/samples_reports/show.haml:293`:
```
// Open the download_nih_tables path in a new tab
    var context = getUrlParameter('context');
    var url = '/samples_reports/'+#{@samples_report.id}+'/nih_tables.html?context='+context+"";
    window.open(url, '_blank');
```